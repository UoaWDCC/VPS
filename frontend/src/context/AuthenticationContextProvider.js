import React, { useState } from "react";
import { signInWithRedirect } from "firebase/auth";
import { useAuthState } from "react-firebase-hooks/auth";
import AuthenticationContext from "./AuthenticationContext";
import { auth, googleProvider } from "../firebase/firebase";
import { useGetSimplified } from "../hooks/crudHooks";

/**
 * This is a Context Provider made with the React Context API
 * AuthenticationContext grants access to functions and variables related to Firebase login
 */
export default function AuthenticationContextProvider({ children }) {
  const [user, loading, error] = useAuthState(auth);

  /**
   * No idToken is stored in state to ensure the non-expired idToken is always used
   * @returns idToken or null if user is not signed in
   */
  async function getUserIdToken() {
    if (user) {
      const token = await user.getIdToken();
      return token;
    }

    return null;
  }

  function signInUsingGoogle() {
    signInWithRedirect(auth, googleProvider);
  }

  function signOut() {
    auth.signOut();
  }

  /**
   * This function determines the role of the current logged in user.
   * @returns the role of user
   */
  function getRole() {
    const [staffList, setStaffList] = useState();
    useGetSimplified("/api/staff", setStaffList);

    let onStaffList = false;
    if (user && getUserIdToken() != null) {
      if (staffList !== undefined) {
        for (let i = 0; i < staffList.length; i += 1) {
          if (staffList[i].firebaseID === user.uid) {
            onStaffList = true;
          }
        }
      }

      if (onStaffList) {
        return "admin";
      }
      return "user";
    }

    return "nonexistent";
  }

  // creating user object with role property
  const VpsUser = {
    firebaseUserObj: user,
    role: getRole(),
  };

  return (
    <AuthenticationContext.Provider
      value={{
        getUserIdToken,
        loading,
        user,
        error,
        signOut,
        signInUsingGoogle,
        VpsUser,
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}
