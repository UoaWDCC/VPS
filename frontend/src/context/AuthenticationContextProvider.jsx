import { signInWithRedirect, signInWithPopup } from "firebase/auth";
import { useState } from "react";
import { useAuthState } from "react-firebase-hooks/auth";
import { auth, googleProvider } from "../firebase/firebase";
import { useGetSimplified } from "../hooks/crudHooks";
import AuthenticationContext from "./AuthenticationContext";

/**
 * This is a Context Provider made with the React Context API
 * AuthenticationContext grants access to functions and variables related to Firebase login
 */
export default function AuthenticationContextProvider({ children }) {
  const [user, loading, error] = useAuthState(auth);

  /**
   * No idToken is stored in state to ensure the non-expired idToken is always used
   * @returns idToken or null if user is not signed in
   */
  async function getUserIdToken() {
    if (user) {
      const token = await user.getIdToken();
      return token;
    }

    return null;
  }

  function signInUsingGoogle() {
    signInWithPopup(auth, googleProvider);
    // signInWithRedirect(auth, googleProvider);
  }

  function signOut() {
    auth.signOut();
  }

  // getting role from backend
  const [userRole, setUserRole] = useState();
  const userID = user == null ? "null" : user.uid; // this is to avoid null pointer exceptions while confining to hook rules
  useGetSimplified(`/api/staff/${userID}`, setUserRole);

  // creating user object with role property
  const VpsUser = {
    firebaseUserObj: user,
    role: userRole,
  };

  return (
    <AuthenticationContext.Provider
      value={{
        getUserIdToken,
        loading,
        user,
        error,
        signOut,
        signInUsingGoogle,
        VpsUser,
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}
