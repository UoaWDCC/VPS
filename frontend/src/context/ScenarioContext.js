import React from "react";

const ScenarioContext = React.createContext();

export default ScenarioContext;
