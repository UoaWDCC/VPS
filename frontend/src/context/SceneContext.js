import React from "react";

const SceneContext = React.createContext();

export default SceneContext;
