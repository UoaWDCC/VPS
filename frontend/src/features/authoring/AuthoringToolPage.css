body {
  background-color: #222;
}

.actions {
  display: flex;
  gap: 6px;
  padding: 5px;
  color: white;
  align-items: center;
  margin-left: 20px;
}

.topbar {
  background-color: #151515;
  width: 100%;
  z-index: 1;
}

.props {
  display: flex;
  padding: 5px;
  flex-wrap: wrap;
  gap: 6px;
  color: white;
  align-items: center;
  margin-left: 20px;
}

.props > div {
  display: flex;
  color: white;
  align-items: center;
}

.button {
  position: relative;
  width: 28px;
  height: 28px;
  border-radius: 5px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.button:hover {
  background-color: #0f0f0f;
}

.text-input {
  background-color: #000;
  border-radius: 5px;
  padding: 0 10px;
  height: 28px;
}

.color-input {
  position: relative;
  width: 18px;
  height: 18px;
  border-bottom: 3px solid white;
  display: flex;
  justify-content: center;
  align-items: center;
}

.color-input span {
  font-size: 14px;
}

.button.active {
  background-color: #000;
}

input[type="number"]::-webkit-outer-spin-button,
input[type="number"]::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

input[type="number"] {
  -moz-appearance: textfield;
}
