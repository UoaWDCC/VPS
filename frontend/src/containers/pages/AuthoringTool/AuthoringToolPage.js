import React, { useContext, useEffect, useState } from "react";
import { useHistory, useParams } from "react-router-dom";
import { Button } from "@material-ui/core";
import TopBar from "../../../components/TopBar";
import ToolBar from "./ToolBar/ToolBar";
import Canvas from "./Canvas/Canvas";
import CanvasSideBar from "./CanvasSideBar/CanvasSideBar";
import ScreenContainer from "../../../components/ScreenContainer";
import ScenarioContext from "../../../context/ScenarioContext";
import { useGet, usePut } from "../../../hooks/crudHooks";
import SceneContext from "../../../context/SceneContext";
import AuthoringToolContext from "../../../context/AuthoringToolContext";
import ToolbarContextProvider from "../../../context/ToolbarContextProvider";
import AuthenticationContext from "../../../context/AuthenticationContext";
import { uploadFiles } from "../../../firebase/storage";
import HelpButton from "../../../components/HelpButton";
import SceneNavigator from "./SceneNavigator/SceneNavigator";

/**
 * This page allows the user to edit a scene.
 * @container
 */
export default function AuthoringToolPage() {
  const { scenarioId, sceneId } = useParams();
  const {
    scenes,
    currentScene,
    setCurrentScene,
    setMonitorChange,
    setHasChange,
    reFetch,
  } = useContext(SceneContext);
  const { currentScenario } = useContext(ScenarioContext);
  const { setSelect } = useContext(AuthoringToolContext);
  const { getUserIdToken } = useContext(AuthenticationContext);
  const [firstTimeRender, setFirstTimeRender] = useState(true);
  const [saveButtonText, setSaveButtonText] = useState("Save");
  const [disableNext, setDisableNext] = useState(false);
  const [disableBefore, setDisableBefore] = useState(false);
  const history = useHistory();
  useGet(
    `/api/scenario/${currentScenario?._id}/scene/full/${currentScene?._id}`,
    setCurrentScene,
    false
  );

  useEffect(async () => {
    reFetch();
  }, []);

  useEffect(() => {
    if (firstTimeRender) {
      setFirstTimeRender(false);
    } else {
      setMonitorChange(true);
    }
  }, [currentScene]);

  /** used to save the scene, as a helper function */
  async function saveScene() {
    setSelect(null);
    await uploadFiles(
      currentScene?.components,
      currentScenario._id,
      currentScene._id
    );
    await usePut(
      `/api/scenario/${scenarioId}/scene/${sceneId}`,
      {
        name: currentScene.name,
        components: currentScene?.components,
      },
      getUserIdToken
    );
    setHasChange(false);
    reFetch();
  }

  /** called when save button is clicked */
  async function save() {
    saveScene();
    setSaveButtonText("Saved!");
    setTimeout(() => {
      setSaveButtonText("Save");
    }, 1800);
  }

  /** called when save and close button is clicked */
  function savePlusClose() {
    saveScene();
    /* redirects user to the scenario page */
    window.location.href = `/scenario/${currentScenario?._id}`;
  }

  function updateScene(direction) {
    const ids = scenes.map((scene) => scene._id);
    const oldIndex = ids.indexOf(currentScene._id);
    const newIndex = oldIndex + direction;

    if (newIndex > -1 && newIndex < ids.length) {
      saveScene();
      setDisableBefore(false);
      setDisableNext(false);
      setCurrentScene(scenes[newIndex]);
      history.push({
        pathname: `/scenario/${scenarioId}/scene/${scenes[newIndex]._id}`,
      });
    }

    if (newIndex === ids.length - 1) {
      setDisableNext(true);
    }

    if (newIndex === 0) {
      setDisableBefore(true);
    }
  }

  return (
    <>
      <ScreenContainer vertical>
        <TopBar back={`/scenario/${currentScenario?._id}`} confirmModal>
          <Button
            className="btn top contained white"
            color="default"
            variant="contained"
            onClick={save}
          >
            {saveButtonText}
          </Button>
          <Button
            className="btn top contained white"
            color="default"
            variant="contained"
            onClick={savePlusClose}
          >
            Save & close
          </Button>
          <Button
            className={`btn top contained white ${
              disableBefore ? "disabled" : ""
            }`}
            color="default"
            variant="contained"
            disabled={disableBefore}
            onClick={() => updateScene(-1)}
          >
            Before
          </Button>
          <Button
            className={`btn top contained white ${
              disableNext ? "disabled" : ""
            }`}
            color="default"
            variant="contained"
            disabled={disableNext}
            onClick={() => updateScene(1)}
          >
            Next
          </Button>
          <HelpButton />
        </TopBar>
        <ToolbarContextProvider>
          <ToolBar />
        </ToolbarContextProvider>
        <div className="flex" style={{ height: "100%", overflow: "hidden" }}>
          <SceneNavigator saveScene={saveScene} />
          <Canvas />
          <CanvasSideBar />
        </div>
      </ScreenContainer>
    </>
  );
}
