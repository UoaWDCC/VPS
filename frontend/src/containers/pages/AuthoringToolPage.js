import React, { useContext } from "react";
import TopBar from "../../components/TopBar";
import ScreenContainer from "../../components/ScreenContainer";
import SceneContext from "../../context/SceneContext";
import ScenarioContext from "../../context/ScenarioContext";
import { useGet } from "../../hooks/crudHooks";

export default function AuthoringToolPage() {
  const { currentScene, setCurrentScene } = useContext(SceneContext);
  const { currentScenario } = useContext(ScenarioContext);
  useGet(
    `/api/scenario/${currentScenario?._id}/scene/full/${currentScene?._id}`,
    setCurrentScene
  );

  return (
    <>
      <ScreenContainer vertical>
        <TopBar back={`/scenario/${currentScenario?._id}`} />
        <h1>Authoring Tool Page</h1>
      </ScreenContainer>
    </>
  );
}
