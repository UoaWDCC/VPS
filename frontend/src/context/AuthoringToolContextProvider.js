import React, { useState, useContext } from "react";
import AuthoringToolContext from "./AuthoringToolContext";
import SceneContext from "./SceneContext";

export default function AuthoringToolContextProvider({ children }) {
  const { currentScene, setCurrentScene } = useContext(SceneContext);

  const [select, setSelect] = useState(null);
  const [bounds, setBounds] = useState(null);
  const [shiftPressed, setShiftPressed] = useState(false);

  function selectElement({ currentTarget }) {
    const image = currentTarget.firstElementChild?.nodeName === "IMG";
    setSelect(image ? currentTarget.firstElementChild.id : currentTarget.id);
  }

  const clearElement = ({ target }) => {
    if (target.id === "canvas") {
      setSelect(null);
    }
  };

  function deleteElement() {
    if (select !== null) {
      console.log(currentScene.components);
      console.log(select);
      const updatedComponents = currentScene.components;
      console.log(updatedComponents.splice(select, 1));

      console.log(updatedComponents);

      setSelect(null);

      setCurrentScene({
        ...currentScene,
        components: updatedComponents,
      });
    }
  }

  return (
    <AuthoringToolContext.Provider
      value={{
        select,
        setSelect,
        selectElement,
        clearElement,
        bounds,
        setBounds,
        shiftPressed,
        setShiftPressed,
        deleteElement,
      }}
    >
      {children}
    </AuthoringToolContext.Provider>
  );
}
