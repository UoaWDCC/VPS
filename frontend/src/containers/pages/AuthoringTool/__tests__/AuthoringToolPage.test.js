import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import AuthoringToolPage from "../AuthoringToolPage";
import SceneContext from "../../../../context/SceneContext";
import ScenarioContext from "../../../../context/ScenarioContext";
import AuthoringToolContextProvider from "../../../../context/AuthoringToolContextProvider";
import AuthenticationContextProvider from "../../../../context/AuthenticationContextProvider";

test("Authoring Tool page snapshot test", () => {
  const context = {
    currentScenario: { _id: "scenarioId" },
    currentScene: { _id: "sceneId", components: [] },
    setMonitorChange: () => {},
  };

  const { baseElement } = render(
    <BrowserRouter>
      <AuthenticationContextProvider>
        <ScenarioContext.Provider value={context}>
          <SceneContext.Provider value={context}>
            <AuthoringToolContextProvider>
              <AuthoringToolPage />
            </AuthoringToolContextProvider>
          </SceneContext.Provider>
        </ScenarioContext.Provider>
      </AuthenticationContextProvider>
    </BrowserRouter>
  );

  expect(baseElement).toMatchSnapshot();
});
