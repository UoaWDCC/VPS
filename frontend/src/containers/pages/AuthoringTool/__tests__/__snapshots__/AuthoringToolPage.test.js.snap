// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authoring Tool page snapshot test 1`] = `
<body>
  <div>
    <div
      class="colContainer"
    >
      <div
        class="topBar"
      >
        <ul
          class="leftTopBarList"
        >
          <li
            class="listItem"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined btn top outlined white"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Back
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
        </ul>
        <ul
          class="rightTopBarList"
        >
          <li
            class="listItem"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained btn top contained white"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Save
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
        </ul>
      </div>
      <div />
      <ul
        class="MuiList-root toolBar MuiList-padding"
        role="menu"
        tabindex="-1"
      >
        <div
          class="MuiBox-root MuiBox-root-1"
          tabindex="0"
        >
          <div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text menuItem"
              tabindex="0"
              title="Background Image"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M21 19V5c0-1.1-.9-2-2-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zM8.5 13.5l2.5 3.01L14.5 12l4.5 6H5l3.5-4.5z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text menuItem"
              tabindex="0"
              title="Text"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M2.5 4v3h5v12h3V7h5V4h-13zm19 5h-9v3h3v7h3v-7h3V9z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text menuItem"
              tabindex="0"
              title="Audio"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 9v6h4l5 5V4L7 9H3zm13.5 3c0-1.77-1.02-3.29-2.5-4.03v8.05c1.48-.73 2.5-2.25 2.5-4.02zM14 3.23v2.06c2.89.86 5 3.54 5 6.71s-2.11 5.85-5 6.71v2.06c4.01-.91 7-4.49 7-8.77s-2.99-7.86-7-8.77z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </ul>
      <div
        class="flex"
        style="height: 100%;"
      >
        <div
          class="moveable-control-box    rCSuehiy8"
          data-able-draggable="true"
          data-able-origin="true"
          data-able-padding="true"
          data-able-resizable="true"
          data-able-snappable="true"
          data-styled-id="rCSuehiy8"
          style="position: absolute; display: none; transform: translate3d(0px, 0px, 0); --zoom: 1; --zoompx: 1px;"
        >
          <div
            class="moveable-control moveable-direction moveable-n "
            data-direction="n"
            data-rotation="0"
            style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
          />
          <div
            class="moveable-control moveable-direction moveable-w "
            data-direction="w"
            data-rotation="90"
            style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
          />
          <div
            class="moveable-control moveable-direction moveable-s "
            data-direction="s"
            data-rotation="0"
            style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
          />
          <div
            class="moveable-control moveable-direction moveable-e "
            data-direction="e"
            data-rotation="90"
            style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
          />
          <div
            class="moveable-control moveable-direction moveable-nw "
            data-direction="nw"
            data-rotation="135"
            style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
          />
          <div
            class="moveable-control moveable-direction moveable-ne "
            data-direction="ne"
            data-rotation="45"
            style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
          />
          <div
            class="moveable-control moveable-direction moveable-sw "
            data-direction="sw"
            data-rotation="45"
            style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
          />
          <div
            class="moveable-control moveable-direction moveable-se "
            data-direction="se"
            data-rotation="135"
            style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
          />
          <div
            class="moveable-control moveable-origin"
            style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
          />
          <div
            class="moveable-line moveable-direction "
            data-direction=""
            data-line-index="0"
            data-rotation="-1"
            style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"
          />
          <div
            class="moveable-line moveable-direction "
            data-direction=""
            data-line-index="1"
            data-rotation="-1"
            style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"
          />
          <div
            class="moveable-line moveable-direction "
            data-direction=""
            data-line-index="2"
            data-rotation="-1"
            style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"
          />
          <div
            class="moveable-line moveable-direction "
            data-direction=""
            data-line-index="3"
            data-rotation="-1"
            style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"
          />
        </div>
        <div
          class="canvasContainer"
        >
          <div
            class="canvas"
            id="canvas"
          >
            <div
              class="MuiBox-root MuiBox-root-3 btn top  white makeStyles-testButton-2"
              id="0"
            >
              <img
                alt="1"
                src="https://avatars.githubusercontent.com/u/52368549?v=4"
              />
            </div>
            <div
              class="MuiBox-root MuiBox-root-4 btn top  white makeStyles-testButton-2"
              id="1"
            >
              <img
                alt="1"
                src="https://avatars.githubusercontent.com/u/52368549?v=4"
              />
            </div>
          </div>
        </div>
        <div
          class="sideBar"
        >
          <div
            class="sceneSettingsContainer"
          >
            <h1
              class="sideBarHeader"
            >
              Scene Settings
            </h1>
            <div
              class="sideBarBody"
            >
              <div
                class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-5 MuiFormControl-fullWidth"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                  data-shrink="false"
                >
                  Scene Name
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="componentPropertiesContainer"
          >
            <h1
              class="sideBarHeader"
            >
              Component Properties
            </h1>
            <div
              class="sideBarBody empty"
            >
              <p
                class="noComponentSelectedText"
              >
                No component selected
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
